timelock-impl/src/main/java/com/palantir/atlasdb/timelock/auth/impl/CachingAuthenticator.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.auth.impl;
+
+import java.util.Map;
+import java.util.Optional;
+
+import javax.ws.rs.ForbiddenException;
+import javax.ws.rs.NotAuthorizedException;
+
+import org.immutables.value.Value;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.palantir.atlasdb.timelock.auth.api.Authenticator;
+import com.palantir.atlasdb.timelock.auth.api.BCryptedSecret;
+import com.palantir.atlasdb.timelock.auth.api.Client;
+import com.palantir.atlasdb.timelock.auth.api.Password;
+
+public class CachingAuthenticator implements Authenticator {
+    private final Map<String, BCryptedSecret> credentials;
+
+    private final LoadingCache<ClientCredentials, Optional<Client>> cache;
+
+    CachingAuthenticator(Map<String, BCryptedSecret> credentials) {
```

Comment:
I can never decide, but do we want to separate out the caching layer and the actual auth layer? I'm not fussed either way, just wanted to bring it up.